### ISSUE TITLE
Beta-testing 'mini.snippets'

### ISSUE THREAD

**Issue by echasnovski** [echasnovski/mini.nvim#1428]
Please leave your feedback about new [mini.snippets](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md) module here. Feel free to either add new comment or positively upvote existing one.

Some things I am interested to find out (obviously, besides bugs):

- Does it work well out of the box with [recommended "Quickstart" setup](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#quickstart)?
- Is configuration intuitive enough?
- Are default values of settings convenient for you?
- Is documentation and examples clear enough?

Thanks!

**Comment by 231tr0n** [comment]
Yay!! Christmas present from the mini.nvim santa has dropped. As usual awesome work @echasnovski. I had few doubts
1. Are snippets populated in mini.completion? Because, I just tried it in a go file and snippets are not showing up.
2. ```lua
	add("rafamadriz/friendly-snippets")
	require("mini.snippets").setup()
```
Is doing this enough to load snippets from the above plugin. Because, when I try to expand it using C-j, it prints out this error message.
![image](https://github.com/user-attachments/assets/e8a9ea2f-33fc-4a31-8cfc-80aaa7994763)

**Comment by echasnovski** [comment]
> 1. Are snippets populated in mini.completion? Because, I just tried it in a go file and snippets are not showing up.

No, not yet. I hope this to get resolved next (#886).



> 2\. Is running the below snippet enough to load snippets from the above plugin

No, 'mini.snippets' by default doesn't load any snippets. This has to be done explicitly by the user (similar to 'mini.hipatterns'). The setup from [Quickstart](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#quickstart) should work.

**Comment by 231tr0n** [comment]
I more doubt, so I have snippet support enabled with gopls the language server for golang, Are there any mappings which this module provides to navigate between these ${1:} and ${n:}, etc.
```go
	fmt.Println(${1:})
```

**Comment by echasnovski** [comment]
> I more doubt, so I have snippet support enabled with gopls the language server for golang, Are there any mappings which this module provides to navigate between these ${1:} and ${n:}, etc.

Yes, `<C-l>` and `<C-h>`. Please, see [Overview](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#expand).

**Comment by 231tr0n** [comment]
One last ask, is it possible to provide an example on how to configure mini.snippets with rafamadriz/friendly-snippets in the docs.

**Comment by echasnovski** [comment]
> One last ask, is it possible to provide an example on how to configure mini.snippets with rafamadriz/friendly-snippets in the docs.

There is. [Quickstart](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#quickstart).

**Comment by 231tr0n** [comment]
> > One last ask, is it possible to provide an example on how to configure mini.snippets with rafamadriz/friendly-snippets in the docs.
> 
> There is. [Quickstart](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#quickstart).

So, basically gen_loader.from_lang() does the job of loading the snippets right?

**Comment by echasnovski** [comment]
> > > One last ask, is it possible to provide an example on how to configure mini.snippets with rafamadriz/friendly-snippets in the docs.
> > 
> > 
> > There is. [Quickstart](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#quickstart).
> 
> So, basically gen_loader.from_lang() does the job of loading the snippets right?

Correct. The [way it works](https://github.com/echasnovski/mini.nvim/blob/2011aff270bcd3e1f3ad088253ace2d574967bed/doc/mini-snippets.txt#L720) is designed after 'rafamadriz/friendly-snippets'. Plus it is a reasonable way to organize per language snippets anyway.

**Comment by 231tr0n** [comment]
```lua
	require("mini.snippets").setup({
		require("mini.snippets").gen_loader.from_lang(),
	})
```
With this setup, it still throws the above error message.

**Comment by echasnovski** [comment]
> Ok! I got it working but, C-h and C-l do not work or navigate for lsp completed snippets ![image](https://private-user-images.githubusercontent.com/56352048/398193453-f1302f79-fd67-4725-998b-82d111c3a93e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ5NjY4MjYsIm5iZiI6MTczNDk2NjUyNiwicGF0aCI6Ii81NjM1MjA0OC8zOTgxOTM0NTMtZjEzMDJmNzktZmQ2Ny00NzI1LTk5OGItODJkMTExYzNhOTNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjIzVDE1MDg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1NmRiM2Y3ODlmOGZiOWY3YWY1NWIyOTg4OTM3ZWU5ZDMzOWVhOWI4Zjk1MWIzZmZmY2FhYzgwMTNmMWQ3NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.crqUuGA4wo8OcS2Ywyoo9RT9ANmAJrJbqexusRl0rwY)

Yes, they do not. Because #886 is still not resolved.

**Comment by 231tr0n** [comment]
> > Ok! I got it working but, C-h and C-l do not work or navigate for lsp completed snippets ![image](https://private-user-images.githubusercontent.com/56352048/398193453-f1302f79-fd67-4725-998b-82d111c3a93e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ5NjY4MjYsIm5iZiI6MTczNDk2NjUyNiwicGF0aCI6Ii81NjM1MjA0OC8zOTgxOTM0NTMtZjEzMDJmNzktZmQ2Ny00NzI1LTk5OGItODJkMTExYzNhOTNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjIzVDE1MDg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1NmRiM2Y3ODlmOGZiOWY3YWY1NWIyOTg4OTM3ZWU5ZDMzOWVhOWI4Zjk1MWIzZmZmY2FhYzgwMTNmMWQ3NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.crqUuGA4wo8OcS2Ywyoo9RT9ANmAJrJbqexusRl0rwY)
> 
> Yes, they do not. Because #886 is still not resolved.

Ohh ok my bad, did not think these two were dependent for working. Overall, the module looks great. I just tried writing basic boiler plate code and I write it bonkers fast now. Thank you once again @echasnovski .

**Comment by xzbdmw** [comment]
I find <c-n> does not work given the config, it delete all the chars before cursor pos, after I expand and then use <right> to move one char, <c-n> works.
```lua
local root = vim.fn.fnamemodify("./.repro", ":p")
-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
    vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end
-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
    vim.fn.system({
        "git",
        "clone",
        "--filter=blob:none",
        "--single-branch",
        "https://github.com/folke/lazy.nvim.git",
        lazypath,
    })
end
vim.opt.runtimepath:prepend(lazypath)
-- install plugins
local plugins = {
    "folke/tokyonight.nvim",
    -- do not remove the colorscheme!
    {
        "echasnovski/mini.snippets",
        version = false,
        config = function()
            local gen_loader = require("mini.snippets").gen_loader
            require("mini.snippets").setup({
                snippets = {
                    -- Load custom file with global snippets first (adjust for Windows)
                    gen_loader.from_file("~/.config/nvim/snippets/global.json"),

                    -- Load snippets based on current language by reading files from
                    -- "snippets/" subdirectories from 'runtimepath' directories.
                    gen_loader.from_lang(),
                },
                -- Module mappings. Use `''` (empty string) to disable one.
                mappings = {
                    -- Expand snippet at cursor position. Created globally in Insert mode.
                    expand = "<C-7>",

                    -- Interact with default `expand.insert` session.
                    -- Created for the duration of active session(s)
                    jump_next = "<C-n>",
                    jump_prev = "<C-p>",
                    stop = "<C-c>",
                },

                -- Functions describing snippet expansion. If `nil`, default values
                -- are `MiniSnippets.default_<field>()`.
                expand = {
                    -- Resolve raw config snippets at context
                    prepare = nil,
                    -- Match resolved snippets at cursor position
                    match = nil,
                    -- Possibly choose among matched snippets
                    select = nil,
                    -- Insert selected snippet
                    insert = nil,
                },
            })
        end,
    },

    -- add any other pugins here
}

require("lazy").setup(plugins, {
    root = root .. "/plugins",
})
vim.cmd([[colorscheme tokyonight]])
```
global.json
```json
{
	"Basic": {
		"prefix": "ba",
		"body": "T1=$1 T2=$2 T0=$0"
	},
}
```

https://github.com/user-attachments/assets/a5c6f07b-781c-4b9e-af77-9e8d56053b55

**Comment by echasnovski** [comment]
> I find "c-n" does not work given the config, it delete all the chars before cursor pos, after I expand and then use "right" to move one char, "c-n" works.

Yeah, unfortunately, `<C-n>` and `<C-p>` are somewhat special because they are forced by built-in Insert completion. My suggestion would be to not use those keys, as 'mini.snippets' has to deal with built-in completion for tabstops with choices. I've spent too much time overcoming its quirks when it comes to handling some common cases when used with 'mini.completion' (there is a [separate test set](https://github.com/echasnovski/mini.nvim/blob/2011aff270bcd3e1f3ad088253ace2d574967bed/tests/test_snippets.lua#L4148-L4235)).

I'll take a closer to try and tackle this case, but I have a feeling that this will just be a documented known limitation. Thanks for the feedback!

**Comment by xzbdmw** [comment]
I tried to remap `<c-n>` to force using `MiniSnippets.session.jump("next")` but with no luck...

**Comment by echasnovski** [comment]
> I tried to remap `<c-n>` to force using `MiniSnippets.session.jump("next")` but with no luck...

That's already what those mapping do (but only for the duration of snippet session).

The issue here is that 'mini.snippets' sues `vim.fn.complete(..., {})` to force hide completion popup. For certain complicated reasons, it has to be done on every jump regardless on whether popup is shown (or at least it is the best solution I could find). The `vim.fn.complete()` call with empty array of items still acts as if completion is active (which is questionable, I think) and with it both `<C-n>` and `<C-p>` are forcefully act to go down and up completion list. Here is a more or less related issue: vim/vim#6440

You can see how that looks with the following:
- `nvim --clean`.
- `:inoremap <C-n> w`
- Type text `aa bb` and on new line press `<C-x><C-n>`. This starts built-in completion. Now pressing `<C-n>` moves down the completion list. If there is no completion, pressing `<C-n>` just inserts `w`.

At the moment, I think that making `vim.fn.complete(..., {})` actually stop current completion and not start new one is the best way forward. But that requires writing Vim C patch... I'll take a look.

**Comment by xzbdmw** [comment]
Oh, thanks for explanation -- I see many hard coded motions,  there are even hard coded keymaps!

**Comment by xzbdmw** [comment]
Stay in insert mode is superior, no longer need to escape select mode when jump to previous tabstop! I'm looking to write a cmp-minisnippets.

**Comment by xzbdmw** [comment]
I notice one very strange thing when using cmp, some chars are swallowed when expanding:
typing sequence : `vim.sc<cr>fun<cr>` result in right pair missing;
typing sequence : `vim.sc<cr><cr>` has correct  result,
typing sequence : `vim.schedule()<left>fun<cr>` has correct  result,
probably a cmp bug.

https://github.com/user-attachments/assets/3fd330eb-0d10-46eb-9f28-6b7ee499df40

```repro.lua
local root = vim.fn.fnamemodify("./.repro", ":p")
-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
    vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end
-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
    vim.fn.system({
        "git",
        "clone",
        "--filter=blob:none",
        "--single-branch",
        "https://github.com/folke/lazy.nvim.git",
        lazypath,
    })
end
vim.opt.runtimepath:prepend(lazypath)
-- install plugins
local plugins = {
    -- do not remove the colorscheme!
    "folke/tokyonight.nvim",
    "neovim/nvim-lspconfig",
    {
        "hrsh7th/nvim-cmp",
        lazy = false,
        dependencies = {
            "hrsh7th/cmp-nvim-lsp",
        },
        config = function(_, opts)
            local cmp = require("cmp")
            require("cmp").setup({
                mapping = cmp.mapping.preset.insert({
                    ["<cr>"] = cmp.mapping(function(fallback)
                        if cmp.visible() then
                            cmp.confirm()
                        end
                    end, { "i", "c", "s" }),
                }),
                completion = {
                    completeopt = "menu,menuone,noinsert",
                },
                snippet = {
                    expand = function(args)
                        local insert = MiniSnippets.config.expand.insert or MiniSnippets.default_insert
                        insert({ body = args.body })
                    end,
                },
                sources = require("cmp").config.sources({
                    { name = "nvim_lsp" },
                }, {}),
            })
        end,
    },
    {
        "echasnovski/mini.snippets",
        version = false,
        config = function()
            local gen_loader = require("mini.snippets").gen_loader
            require("mini.snippets").setup({
                snippets = {
                    -- Load custom file with global snippets first (adjust for Windows)
                    gen_loader.from_file("~/.config/nvim/snippets/global.json"),

                    -- Load snippets based on current language by reading files from
                    -- "snippets/" subdirectories from 'runtimepath' directories.
                    gen_loader.from_lang(),
                },
            })
        end,
    },
    -- add any other pugins here
}

require("lazy").setup(plugins, {
    root = root .. "/plugins",
})

require("lspconfig").lua_ls.setup({
    settings = {
        capabilities = require("cmp_nvim_lsp").default_capabilities(),
        Lua = {
            runtime = {
                version = "LuaJIT",
            },
            workspace = {
                library = {
                    "/usr/local/share/nvim/runtime",
                },
            },
            completion = {
                callSnippet = "Replace",
            },
        },
    },
})
vim.cmd([[colorscheme tokyonight]])
```

**Comment by echasnovski** [comment]
> I notice one very strange thing when using cmp, some chars are swallowed when expanding:
> typing sequence : `vim.sc<cr>fun<cr>` result in right pair missing;
> typing sequence : `vim.sc<cr><cr>` has correct result,
> typing sequence : `vim.schedule()<left>fun<cr>` has correct result,
> probably a cmp bug. (neovim version 0.10.2, `vim.snippet.expand` and `require("luasnip").lsp_expand(args.body)` works fine)

Hmmm... I have doubts that this is from 'mini.snippets' as it during initial insert it doesn't remove any text. But the fact that `vim.snippet.expand()` and 'LuaSnip' work make me doubt my doubts.

**Comment by xzbdmw** [comment]
I debug to find cmp removing the pair before snippet expanding, needs more investment.

**Comment by echasnovski** [comment]
> ...
> `vim.snippet.expand` works correctly because it stays in select mode, does not trigger completion at all, only after user typing `fun` lua_ls will begin to send correct TextEdits, anyway, by that time placeholder is already gone.
> 
> To my understanding, the fix would be suspending completion request to stop step 2 from firing when first expanding.

Thanks for such a deep investigation! From the looks of it, is it safe to say that the issue is on 'nvim-cmp' side? The reason it gets noticeable with 'mini.snippets' and not `vim.snippet.expand` / 'L3MON4D3/LuaSnip' is because the first doesn't use Select mode, correct?

If that's the case, could you, maybe, create an issue in 'hrsh7th/nvim-cmp' with reproduction steps?

**Comment by echasnovski** [comment]
> I find "c-n" does not work given the config, it delete all the chars before cursor pos, after I expand and then use "right" to move one char, "c-n" works.

So I indeed tried to work around it, but there is no real solution here. For at least two reasons (in addition to [dealing with some quirks](https://github.com/echasnovski/mini.nvim/issues/1428#issuecomment-2559966409)):
- The 'mini.snippets' uses `complete(..., [])` Vimscript call to hide completion. This is currently the best way to (immediately) hide current completions. The downside is that it *still* makes Neovim think that there is built-in completion going on, thus `<C-n>` and `<C-p>` are forced to be treated as "go to next/previous completion item". This doesn't make much sense to me (as completion list is empty) and I've asked around about the possibility of fully stopping completion in that case. Turns out, similar version of this was [already discussed in Vim](https://github.com/vim/vim/issues/8784#issuecomment-907644258) with the resolution "Let's not pollute this function and instead create another one". I *think* making a PR to Vim with new `pumstop()` function is possible, but it won't fix the issue number two.
- If there is tabstop with choices, built-in completion menu is shown automatically. This means that in that case pressing `<C-n>` and `<C-p>` will not jump to next/previous tabstop but go to next/previous completion item. This is a usability problem so much so that I don't think accounting for this case is worth it.

So I indeed opted for a [separate note about not using `<C-n>` and `<C-p>` in mappings](https://github.com/echasnovski/mini.nvim/blob/c78332b4c71ad3c2a09efe6acd0a51283627258f/doc/mini-snippets.txt#L602-L605).

**Comment by abeldekat** [comment]
Hello @echasnovski,

When a snippet is inserted and the user presses `<esc>`, the snippet restarts when entering `insert` mode again.
However, if the user deletes the snippet the last tabstop symbol is not removed and appears in a wrong place.
I know this can be avoided by using `<c-c>` instead of `<esc>`. Pressing `<esc>` is something I will have a hardtime to unlearn...

To reproduce, see the following `repro`:

<details>
<summary>init.lua</summary>

```lua
--[[
Use:
  mkdir ~/.config/repro
  cd ~/.config/repro

  touch init.lua
  add the contents of this file to init.lua
  NVIM_APPNAME=repro nvim init.lua 

Remove:
  rm -rf ~/.local/share/repro ~/.local/state/repro ~/.local/cache/repro
  rm -rf ~/.config/repro
--]]

--[[
Steps to reproduce:
  1. Navigate to one line above the line containing function clone
  2. Insert mode, type "for" and "<c-j>": The "for" snippet is expanded
  3. Type "esc" and "3dd"
  4. Now the final tabstop symbol is visible as the first character on the line
     containing function clone
  5. I cannot remove the symbol. Reloading the buffer also does not remove the symbol
  6. Restart nvim: The symbol is no longer present
--]]

local function clone(path_to_site)
  local mini_path = path_to_site .. "pack/deps/start/mini.nvim"
  if not vim.uv.fs_stat(mini_path) then
    vim.cmd('echo "Installing `mini.nvim`" | redraw')
    local clone_cmd =
      { "git", "clone", "--filter=blob:none", "https://github.com/echasnovski/mini.nvim", mini_path }
    vim.fn.system(clone_cmd)
    vim.cmd("packadd mini.nvim | helptags ALL")
    vim.cmd('echo "Installed `mini.nvim`" | redraw')
  end
end

local path_to_site = vim.fn.stdpath("data") .. "/site/"
clone(path_to_site)
local MiniDeps = require("mini.deps")
MiniDeps.setup({ path = { package = path_to_site } })
local add, now = MiniDeps.add, MiniDeps.now

now(function()
  add("rafamadriz/friendly-snippets")

  vim.cmd("colorscheme randomhue")
  require("mini.basics").setup() -- sensible defaults

  local gen_loader = require("mini.snippets").gen_loader
  require("mini.snippets").setup({
    snippets = {
      gen_loader.from_lang(),
    },
  })
  local mini_pick = require("mini.pick")
  mini_pick.setup()
  vim.ui.select = mini_pick.ui_select
end)
```

</details>

**Comment by echasnovski** [comment]
> When a snippet is inserted and the user presses `<esc>`, the snippet restarts when entering `insert` mode again.

It does not "restart" because snippet session never stopped. Snippet session stops only in two ways: [manually or automatically](https://github.com/echasnovski/mini.nvim/blob/c78332b4c71ad3c2a09efe6acd0a51283627258f/doc/mini-snippets.txt#L1010-L1023). Exiting into Normal mode will stop session *only* if current tabstop is final, otherwise it is treated as an "exit for quick text edit and then go back to editing in snippet session".

> However, if the user deletes the snippet the last tabstop symbol is not removed and appears in a wrong place.

The answer is simple: [don't do that](https://github.com/echasnovski/mini.nvim/blob/c78332b4c71ad3c2a09efe6acd0a51283627258f/doc/mini-snippets.txt#L1030-L1033). I explored the possibility of stopping session whenever any tabstop extmark is invalidated, but that looked impossible because I want users to be able to delete tabstop text completely (i.e. for it to become empty string in order to be visualized as inline virtual text).

> I know this can be avoided by using `<c-c>` instead of `<esc>`. Pressing `<esc>` is something I will have a hardtime to unlearn...

The important thing to understand here is that "stop snippet session" and "exit into Normal mode" are independent actions (if current tabstop is not final). If you want to stop the session whenever you press `<Esc>`, there are at least two ways to do that:
- Use `<Esc>` as `config.mappings.stop`. It will be created only for the duration of active session(s). First press(es) will stop session(s) (and keep Insert mode) and it will exit Normal mode after there is no active session(s).
- Set up `ModeChanged *:n` autocommand creation whenever session starts. With something like this:
    ```lua
    local modechanged_au_id
    local stop_all_sessions = function()
      -- Stop all nested sessions
      while MiniSnippets.session.get() do
        MiniSnippets.session.stop()
      end
      modechanged_au_id = nil
    end
    local setup_stop_in_normal_mode = function()
      if modechanged_au_id ~= nil then return end
      modechanged_au_id =
        vim.api.nvim_create_autocmd('ModeChanged', { pattern = '*:n', once = true, callback = stop_all_sessions })
    end
    vim.api.nvim_create_autocmd('User', { pattern = 'MiniSnippetsSessionStart', callback = setup_stop_in_normal_mode })
    ```

**Comment by abeldekat** [comment]
Thank you! As always, your answer provides a better understanding of the plugin.

I really like mini.snippets!

In my [config](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/lua/ak/deps/coding.lua#L77) the [plugin](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/lua/ak/config/coding/mini_snippets.lua#L23) has been integrated with [nvim-cmp](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/lua/ak/config/coding/nvim_cmp.lua#L64) to expand `lsp` snippets.
I have one problem though: I already use `<c-j>` for `cmp.mapping.confirm({ select = true })` because typing `<c-y>` is more cumbersome to me.

For the moment I use `mappings = { expand = "<C-S>" }` but this is also not ideal due to a restriction in my [left](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/README.md?plain=1#L55) hand. If you have any advice that would be greatly appreciated.

**Comment by echasnovski** [comment]
> For the moment I use `mappings = { expand = "<C-S>" }` but this is also not ideal due to a restriction in my [left](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/README.md?plain=1#L55) hand. If you have any advice that would be greatly appreciated.

Here are several options:
- Use `<CR>` as "confirm completion item selection" mapping. This is fairly common, I'd guess.
- Use the same key in "expand or jump" style. Similar to [this example](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L407). I've personally grown to not like that (I used `<C-l>` to "expand or jump next") in favor of keeping those separate, but maybe you'll like it.
- The `<C-]>` is also a reasonable key to expand snippet, but it is even more cumbersome to type.
- If you don't want the default behavior of `<C-k>` in Insert mode, you can try it for expanding. And maybe remapping some other key so serve as `<C-k>` in Insert mode.

**Comment by krovuxdev** [comment]
Hi,  
Thank you for your work on `mini.snippet`. I have a suggestion to improve the usability of the plugin.  

Would it be possible to add a keymap, such as `expand_choice = <C-E>`, to allow users to select from predefined options in a choice snippet? This could be similar to how [Choice](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_choice) works in Visual Studio Code snippets.  

This functionality would streamline workflows for users who rely on choice-based snippets.

**Comment by echasnovski** [comment]
> Would it be possible to add a keymap, such as `expand_choice = <C-E>`, to allow users to select from predefined options in a choice snippet? This could be similar to how [Choice](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_choice) works in Visual Studio Code snippets.
> 
> This functionality would streamline workflows for users who rely on choice-based snippets

Choices are [shown using with built-in completion popup](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L225-L229). This means that navigating through completion popup can be done with `<C-n>` ("next") and `<C-p>` ("previous"). Using arrows ("Down" / "Up") and then `<CR>` also works.

Does this answer your question or did I misunderstand something?

**Comment by abeldekat** [comment]
>For the moment I use mappings = { expand = "<C-S>" } but this is also not ideal due to a restriction in my [left](https://github.com/abeldekat/nvim_pde/blob/d5d3b51bcbf60bee36c24d1398a097a384fa9fc8/README.md?plain=1#L55) hand. If you have any advice that would be greatly appreciated.

Thanks for taking the time to answer [this](https://github.com/echasnovski/mini.nvim/issues/1428#issuecomment-2563996818) question! I changed `mappings.expand` into `<c-k>` and wrote the motivation [here.](https://github.com/abeldekat/nvim_pde/blob/645d7e67c04d4ec1c154f04d2ab6cf4186106cae/lua/ak/config/coding/mini_snippets.lua#L23)

Visualizing the tabstops is a great and distinctive feature! As a consequence, I discovered the power of having the `lsp` expand a function call with argument stops ("callSnippet" in "lua_ls"). I can use `<c-l>` during any function completion to navigate to the end of the function call. That behavior is similar to for example [tabout.nvim](https://github.com/abecodes/tabout.nvim]) and a real improvement when appending a ";" at the end of a function call in for example `rust`. No more `<esc>A;` or the like... I am thinking about posting to `reddit` in order to share this "discovery". Would that be useful in your opinion?

Regarding this [comment](https://github.com/echasnovski/mini.nvim/issues/1428#issuecomment-2563668092):

>Yes, you can: enter Insert mode and press <C-c> to explicitly stop session.

As I understand from the help a session can be stopped in two ways:

- press ctrl-c
- navigate to final tabstop, press `esc` or make an edit

Playing around, I think there is a third way that I did not discover when reading the help: Press `<esc>` in the current snippet, move away and start a new snippet. Is it true that there can be at most one "in progress" snippet session?

**Comment by echasnovski** [comment]
> Playing around, I think there is a third way that I did not discover when reading the help: Press `<esc>` in the current snippet, move away and start a new snippet. Is it true that there can be at most one "in progress" snippet session?

No, sessions can be [nested](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L988-L1000) and how it is done is one of the distinctive features of 'mini.snippets' for reasons [described here](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L1002-L1006). The effect is also visible in [the demo](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#demo).

**Comment by abeldekat** [comment]
> I don't think it will hurt
...
The difference with 'mini.snippets' is that the final tabstop is also visualized which has a better usability, I think.

Indeed. As a more or less experienced "vimmer" I never thought about this approach and only figured it out when I saw that final tabstop symbol everywhere I expand a function offered by `lsp`. Suddenly it "clicked", so thank you! Might I suggest to add the approach somewhere in the help for new users?

> No, sessions can be [nested](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L988-L1000) and how it is done is one of the distinctive features of 'mini.snippets' for reasons [described here](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L1002-L1006). The effect is also visible in [the demo](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#demo).

Yes, I did not formulate the question and statement correctly. I am referring to the following situation:

1. In a `lua` file, type `req<c-j` and select the snippet. Type the module name inside the parenthesis and press `esc`
2. Notice that the final tabstop symbol is still visible
3. Move outside of the snippet and repeat step 1 on a new line.
4. Notice that the final tabstop symbol is only visible in the second snippet.

In `lua` statements don't require a delimiter, as such jumping to the end of the `"require("something")` is mostly not needed. I imagine I can have lots of those "require" statements(or any other unfinished snippets due to function expansion) without ever needing to press `<c-c>`. Are there drawbacks to this approach, effectively having multiple open snippet sessions?

**Comment by echasnovski** [comment]
> 4\. Notice that the final tabstop symbol is only visible in the second snippet.

Yes, because only a single session is being visible. If you press `<C-c>` (or stop session in any other way), the previous session will be active and visible. Stopping it again will remove its visualization also.

> In `lua` statements don't require a delimiter, as such jumping to the end of the `"require("something")` is mostly not needed. I imagine I can have lots of those "require" statements(or any other unfinished snippets due to function expansion) without ever needing to press `<c-c>`.

The always present final tabstop is there to always have a way to automatically end session (i.e. without explicit `<C-c>`). So in this case you are expected to press `<C-l>` again making final tabstop current and placing cursor at its final position (i.e. the one designated for when the session is stopped). After that you can either `<Esc>` to make Normal mode navigation or just continue typing, like creating new line.

In this case what I'd recommend is:
- `req<C-j>` - `mini.snippets` - `<C-l>`.
- If you want to keep typing - type. Like `<CR>` will autostop current session and start a new line. If want to stop inserting text, exit into Normal mode, this will also autostop current session.

> Are there drawbacks to this approach, effectively having multiple open snippet sessions?
Not closing previous session will have at least the following side effects:
- Visualization will be present and will be updated according to other text edits. This might lead to confusing outcome, like you described with `dd` case earlier.
- In-session mappings (`<C-h>`, `<C-l>`, `<C-c>`) will be present.

So my recommendation is to follow [Session life cycle](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L965) which recommends session to be stopped when you are confident that you won't need it. The default autostopping rules are there for the most expected design: stop if final is current, but allow going back to fix the error by not doing it immediately.

**Comment by abeldekat** [comment]
> In this case what I'd recommend is:
>
>   `req<C-j>` - mini.snippets - `<C-l>`.
    If you want to keep typing - type. Like <CR> will autostop current session and start a new line. If want to stop inserting text, exit into Normal mode, this will also autostop current session.

I will certainly follow that recommendation. In hindsight I think I did not grasp the implications of the words "session stack approach" in the [session life cycle](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L1002). Thanks for taking the time to explain!

**Comment by krovuxdev** [comment]
> > Would it be possible to add a keymap, such as `expand_choice = <C-E>`, to allow users to select from predefined options in a choice snippet? This could be similar to how [Choice](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_choice) works in Visual Studio Code snippets.
> > This functionality would streamline workflows for users who rely on choice-based snippets
> 
> Choices are [shown using with built-in completion popup](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L225-L229). This means that navigating through completion popup can be done with `<C-n>` ("next") and `<C-p>` ("previous"). Using arrows ("Down" / "Up") and then `<CR>` also works.
> 
> Does this answer your question or did I misunderstand something?



You are correct, but when I use `<C-p>` or `<C-n>`, it shows Neovim's completion menu instead of the choice menu. This happens when I have the following configuration:  

```lua
vim.o.complete = "."
vim.o.completeopt = "menu,menuone,noinsert"
```  

If I change it to:  

```lua
vim.o.complete = ""
vim.o.completeopt = ""
```  

then the choice menu works as expected without showing the Neovim completion menu.  

However, if I use `jump_next` and then `jump_prev`, the choice menu no longer works with `<C-p>` or `<C-n>`. On the other hand, if I delete some text, the choice menu works again with `<C-p>`.  

For example, in `luasnip`, using `<C-E>` allows me to switch between choices without opening the completion menu, even with the following configuration:  

```lua
vim.o.complete = "."
vim.o.completeopt = "menu,menuone,noinsert"
```  

That's what I was referring to! :)  

Thanks for taking the time to explain!

**Comment by echasnovski** [comment]
> You are correct, but when I use `<C-p>` or `<C-n>`, it shows Neovim's completion menu instead of the choice menu. This happens when I have the following configuration:

There is certainly some kind of misunderstanding here.

What do you mean by "choice menu"? 'mini.snippets' shows choices automatically via Neovim's completion menu. Here is what it looks like:

https://github.com/user-attachments/assets/e2053276-158a-4337-a27a-86720d67b7e5

This is with `'menuone,noselect,fuzzy'` as 'completeopt'.

Do you use autocompletion plugin? Could you, please, record the screencast of what do you see as a problem here?

**Comment by krovuxdev** [comment]
> screencast



> > You are correct, but when I use `<C-p>` or `<C-n>`, it shows Neovim's completion menu instead of the choice menu. This happens when I have the following configuration:
> 
> There is certainly some kind of misunderstanding here.
> 
> What do you mean by "choice menu"? 'mini.snippets' shows choices automatically via Neovim's completion menu. Here is what it looks like:
> demo-snippets_choices.mp4
> 
> This is with `'menuone,noselect,fuzzy'` as 'completeopt'.
> 
> Do you use autocompletion plugin? Could you, please, record the screencast of what do you see as a problem here?

https://github.com/user-attachments/assets/6202c349-0112-4302-861a-8d647703842a

**Comment by echasnovski** [comment]
Ahhh, that's what you're talking about. This is currently by design: when already edited tabstop and jump back to it, [only "relevant" choices are shown](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L1111-L1114). That is, only those which match the currently typed text. To show all choices, delete the whole tabstop text and it will "restart" choice suggestion to again show all of them. I realize that this is not properly documented, though.

The other value of 'completeopt' shows the same behavior but without actually showing the popup menu.

Initially it was to account for the fact that user can type some "partial" text and go back later to it. But now I have doubts about if it is a good idea. I'll investigate if changing it to always show all choices makes more sense or making a more clear documentation is better here.

**Comment by abeldekat** [comment]
In my config, [nvim-cmp](https://github.com/abeldekat/nvim_pde/blob/876bac240df8246260dfa9657e4a15435862aec7/lua/ak/config/coding/nvim_cmp.lua#L77) uses `mini.snippets` to expand lsp snippets. 

When I type one letter `nvim-cmp` activates. I can activate `mini.snippets` manually by pressing `<c-j>`, but this does not close the `nvim-cmp` window. As a result, the buffer can have 2 popups(`nvim-cmp` and `vim.ui.select`) that may overlap.

Initially I tweaked the [config](https://github.com/abeldekat/nvim_pde/blob/645d7e67c04d4ec1c154f04d2ab6cf4186106cae/lua/ak/config/coding/mini_snippets.lua#L36) of `mini.snippets` to close `nvim-cmp`.

I would like to do the above in the [config](https://github.com/abeldekat/nvim_pde/blob/876bac240df8246260dfa9657e4a15435862aec7/lua/ak/config/coding/nvim_cmp.lua#L77) of `nvim-cmp` because that's where other resources are connected. The config of `mini.snippets` should ideally not have to know about the presence of `nvim-cmp`as it operates standalone.

Currently this is not possible because event `MiniSnippetsSessionStart` fires "after". Would you consider adding event `MiniSnippetsSessionStartPre` for smoother integration with other resources?

In my [nvim-cmp](https://github.com/abeldekat/nvim_pde/blob/876bac240df8246260dfa9657e4a15435862aec7/lua/ak/config/coding/nvim_cmp.lua#L77) config I now override the public expand method, which is less clean:

```lua
-- mini.snippets: Close cmp:
if snip_engine == "mini" then -- event "MiniSnippetsSessionStart" is too late...
  local org_expand = MiniSnippets.expand
  ---@diagnostic disable-next-line: duplicate-set-field
  MiniSnippets.expand = function(opts)
    if cmp.visible() then cmp.close() end
    org_expand(opts)
  end
end
```

**Comment by echasnovski** [comment]
> Currently this is not possible because event `MiniSnippetsSessionStart` fires "after". Would you consider adding event `MiniSnippetsSessionStartPre` for smoother integration with other resources?
> 
> In my [nvim-cmp](https://github.com/abeldekat/nvim_pde/blob/876bac240df8246260dfa9657e4a15435862aec7/lua/ak/config/coding/nvim_cmp.lua#L77) config I now override the public expand method, which is less clean:

No need to monkey patch public methods. Users have [more granular control](https://github.com/echasnovski/mini.nvim/blob/2faada1cffce5609d78ac5e81eaf3754e0476758/doc/mini-snippets.txt#L607-L664) over how expand is done in each of four steps. As the issue seems to be that `vim.ui.select` and 'nvim-cmp' windows overlap, having `if cmp.visible() then cmp.close() end` in `select` step might be best. So something like this:

```lua
require('mini.snippets').setup({
  -- ...
  expand = {
    select = function(...)
      if cmp.visible() then cmp.close() end
      return MiniSnippets.default_select(...)
    end
  }
})
```

**Comment by abeldekat** [comment]
FYI:
[LazyVim PR](https://github.com/LazyVim/LazyVim/pull/5274)

**Comment by echasnovski** [comment]
> FYI: [LazyVim PR](https://github.com/LazyVim/LazyVim/pull/5274)

Thanks for the effort! Left couple of review comments which I think are relevant.

**Comment by abeldekat** [comment]
In the help, regarding "supertab", you write:

> This module intentionally by default uses separate keys to expand and jump as
> it enables cleaner use of nested sessions. Here is an example of setting up
> custom <Tab> to "expand or jump" and <S-Tab> to "jump to previous": >lua

The word "cleaner" triggered me to think about concrete drawbacks.

In my [config](https://github.com/abeldekat/nvim_pde/blob/a61585f4413a48e3603604e2532c051a3b7d8060/lua/ak/deps/coding.lua#L6) I can combine combinations of `nvim-cmp`, `blink.cmp`, `luasnip`, `mini.snippets` and `vim.snippets`.
I also added "supertab" to the [config](https://github.com/abeldekat/nvim_pde/blob/a307805e30fc01126221bade5404e4f34586b73b/lua/ak/config/coding/mini_snippets.lua#L13)

Apart from not being ubiqitous and thus a little harder to navigate, I found one concrete drawback: Pressing tab at the start of the line or inside an empty tabstop inserts a real "tab" instead of opening the somewhat expected `vim.ui.select` containing all relevant snippets.
Are there other drawbacks? Would it be a good idea to materialize them in the help?

The suggested implementation of supertab in the help:
```lua
  -- ...
  local snippets = require('mini.snippets')
  local match_strict = function(snippets)
    -- Do not match with whitespace to cursor's left
    return snippets.default_match(snippets, { pattern_fuzzy = '%S+' })
  end
  -- ...
```

This introduces warning: "Redefined local snippets" and error `E5108: Error executing lua: .../.config/nvim/lua/ak/config/coding/mini_snippets.lua:53: attempt to call field 'default_match' (a nil value)`

I would suggest the following:

```lua
  -- ...
  local snippets = require('mini.snippets')
  local match_strict = function(candidate_snippets)
    -- Do not match with whitespace to cursor's left
    return snippets.default_match(candidate_snippets, { pattern_fuzzy = '%S+' })
  end
  -- ...
```

Would you be willing to accept a PR adding the candidate_snippets fragment to the help?

**Comment by echasnovski** [comment]
> Pressing tab at the start of the line or inside an empty tabstop inserts a real "tab" instead of opening the somewhat expected `vim.ui.select` containing all relevant snippets.
> Are there other drawbacks? Would it be a good idea to materialize them in the help?

This is a result `pattern_fuzzy = '%S+'` and was deliberate because I'd expect that pressing "tab" after a whitespace (or at the start of line) to insert a real "tab". Otherwise the key is pretty much reserved for snippets and inserting a real "tab" requires `<C-v><Tab>`, which I don't think ideal. Removing `, { pattern_fuzzy = '%S+' }` altogether in the user's code should fix this.

> Would you be willing to accept a PR adding the candidate_snippets fragment to the help?

That's my oversight of not sticking *exactly* to the example in the tests. I am currently working on the module, so will fix it myself (with you as co-author, of course :) ). Thanks for noticing!

**Comment by echasnovski** [comment]
> I intend to write a cmp resource: cmp_mini_snippets.

That's great! Let me know if you need any clarifications and/or 'mini.snippets' adjustment. It would also be best to possibly move that discussion somewhere else. Do you plan to have separate repo for that from the start? If yes, then maybe there; if not, then in separate discussion here.

> While investigating I saw that `LuaSnip` supports "autosnippets". Perhaps this feature could be added to the comparisons section of the help?

Ah, you mean `enable_autosnippets` config? Yeah, didn't notice that. Makes sense to add to comparison. I'll do it today.

**Comment by xzbdmw** [comment]
> Stay in insert mode is superior, no longer need to escape select mode when jump to previous tabstop! I'm looking to write a cmp-minisnippets.

I've made a [simple one](https://github.com/xzbdmw/cmp_mini_snippets), mainly to integrate my [pr](https://github.com/hrsh7th/nvim-cmp/pull/1955)

It looks like this

https://github.com/user-attachments/assets/cd6bed40-0b86-468b-871a-81d472c9b2c4

**Comment by echasnovski** [comment]
> Initially it was to account for the fact that user can type some "partial" text and go back later to it. But now I have doubts about if it is a good idea. I'll investigate if changing it to always show all choices makes more sense or making a more clear documentation is better here.

@krovuxdev, I believe the issue should now be resolved: all choices are shown after jumping and deleting all tabstop text. There is also now [more info about interaction with choices](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L991-L997).

**Comment by krovuxdev** [comment]
Hi @echasnovski,

Thank you very much! It worked, but I noticed a small problem.  
When I don't want to show the menu while expanding the snippet and press `<C-n>`, the menu still appears. This shouldn't happen. I showed the issue in the attached video.  

https://github.com/user-attachments/assets/68fc24bb-ec63-48a3-a7b2-3b262272e479

**Comment by echasnovski** [comment]
> When I don't want to show the menu while expanding the snippet and press `<C-n>`, the menu still appears. This shouldn't happen. I showed the issue in the attached video.

As far as I can tell, the menu appears only after the initial insert and all other times it works as expected, correct? The reason for that is because [`MiniSnippetsSessionStart` is triggered right after a session is started](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L1050) which is a common way to trigger any `Enter`/`Start` type of events. This also means that choices for the first tabstop are already shown at the point that event is triggered. If you want the menu not shown even in this situation, then moving the logic of adjusting 'completeopt' into custom `config.expand.insert()` (i.e. before calling `MiniSnippets.default_insert()`) is the way to go here.

Other than that, the general direction of not showing completion menu by adjusting 'completeopt' is indeed the suggested approach here. Couple of questions/suggestions, though:
- The choices suggestions work best [when there is a `noselect`](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L994) flag there. So I'd change the `nomenu` case to 'noselect'. But it should still work with other 'completeopt' values, just won't be tested for them.
- I am not sure why third type of adjustment in `MiniSnippetsSessionJump` is needed. Was there a particular reason for it to be added?

**Comment by 9seconds** [comment]
@echasnovski this is a first snippet plugin in any editor I've tried that works exactly how I want: sometimes during snippet session I have to check something, like open vertical split etc, and continue with snippet after. Very huge thanks for that!

I have only one suggestion: the way of adding custom variables is not very obvious and documented. This is what I'm doing now (with lazy.nvim):

```lua
{
  opts = function()
    local plugin = require("mini.snippets")

    return {
      expand = {
        insert = function(snippet, opts)
          return plugin.default_insert(snippet, vim.tbl_deep_extend("force", opts or {}, {
            lookup = {
              ["GIT_USERNAME"] = vim.fn.system("git config user.name")
            }
          }))
        end
      }
    }
  end
}
```

This works. But I guess, such case would be very frequent so ideally, I would like to have nice shortcuts just to insert such variables. Do you think it makes sense to expose it to configuration?

**Comment by echasnovski** [comment]
> This works. But I guess, such case would be very frequent so ideally, I would like to have nice shortcuts just to insert such variables. Do you think it makes sense to expose it to configuration?

Adding an example with exactly this approach, which is indeed a suggested one, to variable lookup is planned (didn't realize I missed it).

Regarding this exact `GIT_USERNAME` use case I'd suggest creating it as an environment variable in the current *Neovim* session, like somewhere in the 'init.lua'. Environment variables [are recognized during snippet parsing and normalization](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L247-L248). Does it count as a "nice shortcut"?

**Comment by echasnovski** [comment]
> I just wanted to make some simple illustration without huge piles of specific code

I don't think it is a huge pile. The `vim.tbl_deep_extend` is not needed here, as `config.expand.insert` accepts only a `snippet` as argument. The `opts` table is an argument for `default_insert()` exactly for this occasion of tweaking its options.

So it would be something like this:

```lua
-- Use evnironment variables with value is same for all snippet sessions
vim.loop.os_setenv('USERNAME', 'user')

-- Compute custom lookup for variables with dynamic values
local insert_with_lookup = function(snippet)
  local lookup = { TM_SELECTED_TEXT = vim.fn.getreg('*') }
  return MiniSnippets.default_insert(snippet, { lookup = lookup })
end

require('mini.snippets').setup({
  -- ... Set up snippets ...
  expand = { insert = insert_with_lookup },
})
```

**Comment by 9seconds** [comment]
Seems I've found a bug. Consider a following snippet for Python:

```lua
return {
  ["'try/except' clause"] = {
    prefix = "te",
    body = {
      "try:",
      "\t${0:$TM_SELECTED_TEXT}",
      "except Exception:",
      "\tpass",
    },
  },
}
```

![CleanShot 2025-01-03 at 14 22 26](https://github.com/user-attachments/assets/765d5cb6-439e-426e-bb7d-093db3e1a6af)

The problem is that in case of multiline `$TM_SELECTED_TEXT` only first line is correctly indented so I have to reindent.

I've seen something relevant here: https://github.com/hrsh7th/vim-vsnip/issues/86 Hope it helps :pray:

**Comment by echasnovski** [comment]
> The problem is that in case of multiline `$TM_SELECTED_TEXT` only first line is correctly indented so I have to reindent.

Hmmm... My initial impression is that the current behavior is correct. Resolving variables is best understood as a straightforward replace. In this case it is replaced with `'multi\nline\ntext'` which should be treated as if it was used in the snippet to begin with. I do get that it might be useful the other way around, but still.

Tweaking indent during variable expansion also opens a big Pandora box. For example, the case of several such variables together. Like in case of `'\t$TM_SELECTED_TEXT$TM_SELECTED_TEXT'` where selected text itself adds indent at its last line.

The other technical reason is that indents in 'mini.snippets' respect comment leaders which might introduce problems.

I'll think about the concise way to possibly adjust for that, but at the moment I am a bit skeptical.

**Comment by radmen** [comment]
Hello. Thank you very much for the plugin. I think this is exactly what I wanted.

I tried to scoop the thread and see if this was mentioned something. Apologies if I duplicated the reports.

1. is there a quick way to remove/blank the placeholder value? Sometimes I add them as hints, but would like to be able to quickly remove them if I want to keep the value empty. Currently, I handle this by `<Space><Backspace>`
2. Would it be possible to highlight the placeholder? I have a snippet with a nested variable and would like to see how "far" it reaches (mostly to be able to remove it if I don't need it).

Here is a short demo vid:

https://github.com/user-attachments/assets/919a4040-3122-46db-ba52-fc8850691d8e

And the snippet used in the video:

```json
  "function": {
    "prefix": "fun",
    "body": [
      "function (${2:arg})${3: use ($$4)} {",
      "\t$0",
      "}"
    ]
  },
```

PS: I just realized that I could just add a second snippet just for the (`use ($$4))`) part, though I still would find it useful if the things I mentioned are possible.

Thank you!

**Comment by echasnovski** [comment]
> Hello. Thank you very much for the plugin. I think this is exactly what I wanted.

:tada:

>     1. is there a quick way to remove/blank the placeholder value? Sometimes I add them as hints, but would like to be able to quickly remove them if I want to keep the value empty. Currently, I handle this by `<Space><Backspace>`

I also encountered it, but I am afraid the current approach of "remove placeholder after typing at its start" is the best design here. Mostly because placeholders (in my opinion) should contain "valid" text in the context of the snippet (for programming languages - something that syntactically correct, for example). Or more strictly they are "leave it as is or replace" type of text. So using `<Space><BS>` is the way here.

>     2. Would it be possible to highlight the placeholder? I have a snippet with a nested variable and would like to see how "far" it reaches (mostly to be able to remove it if I don't need it).

It is highlighted. Depending on the tabstop and session progression, tabstops are highlighted with one of [five highlight groups](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L117-L121). Those by default are colored underdouble which is not really seen in your screencast. Check out [demo](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md#demo) to see how they can be seen. Or define are more visible highlight groups for your terminal/color scheme. Here is a crude quick example:

```lua
vim.api.nvim_set_hl(0, 'MiniSnippetsCurrent', { bg = 'Yellow' })
vim.api.nvim_set_hl(0, 'MiniSnippetsCurrentReplace', { bg = 'Red' })
vim.api.nvim_set_hl(0, 'MiniSnippetsFinal', { bg = 'Green' })
vim.api.nvim_set_hl(0, 'MiniSnippetsUnvisited', { bg = 'Cyan' })
vim.api.nvim_set_hl(0, 'MiniSnippetsVisited', { bg = 'Blue' })
```

**Comment by radmen** [comment]
> Mostly because placeholders (in my opinion) should contain "valid" text in the context of the snippet

Yeah, I think you're right on this one. It hit me once I started thinking in possible workarounds for my issues.

> It is highlighted. Depending on the tabstop and session progression, tabstops are highlighted with one of [five highlight groups](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L117-L121)

Oh, got it. So it's more an issue of my color theme.

Thank you!

**Comment by echasnovski** [comment]
> > The problem is that in case of multiline `$TM_SELECTED_TEXT` only first line is correctly indented so I have to reindent.
> 
> Hmmm... My initial impression is that the current behavior is correct. Resolving variables is best understood as a straightforward replace. In this case it is replaced with `'multi\nline\ntext'` which should be treated as if it was used in the snippet to begin with. I do get that it might be useful the other way around, but still.

@9seconds, after quick testing I agree that having some kind of indent adjustments in case of expanding variables is more expected. Probably, nested placeholders also (like in `$1\n${2:\t$1}` or `$1\n\t${2:$1}`). I'll have to do a research and think about how to reasonably achieve this.

> I have only one suggestion: the way of adding custom variables is not very obvious and documented. This is what I'm doing now (with lazy.nvim):

Also, this is now [documented](https://github.com/echasnovski/mini.nvim/blob/546b9c8bd8ebba594bd18811e54bd3e6f9a662ec/doc/mini-snippets.txt#L448-L465).

**Comment by abeldekat** [comment]
I submitted a [PR](https://github.com/Saghen/blink.cmp/pull/877)  in blink.cmp.

For caching, I saw the context you mentioned earlier. I copied `H.get_default_context` and modified the return value to be only `lang`. Is there a reason you are using the `buf_id/lang` combination?

**Comment by krovuxdev** [comment]
> > When I don't want to show the menu while expanding the snippet and press `<C-n>`, the menu still appears. This shouldn't happen. I showed the issue in the attached video.
> 
> As far as I can tell, the menu appears only after the initial insert and all other times it works as expected, correct? The reason for that is because [`MiniSnippetsSessionStart` is triggered right after a session is started](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L1050) which is a common way to trigger any `Enter`/`Start` type of events. This also means that choices for the first tabstop are already shown at the point that event is triggered. If you want the menu not shown even in this situation, then moving the logic of adjusting 'completeopt' into custom `config.expand.insert()` (i.e. before calling `MiniSnippets.default_insert()`) is the way to go here.
> 
> Other than that, the general direction of not showing completion menu by adjusting 'completeopt' is indeed the suggested approach here. Couple of questions/suggestions, though:
> 
>     * The choices suggestions work best [when there is a `noselect`](https://github.com/echasnovski/mini.nvim/blob/782969a40d2870b1d7cd7c9db8f965d8aa9705ef/doc/mini-snippets.txt#L994) flag there. So I'd change the `nomenu` case to 'noselect'. But it should still work with other 'completeopt' values, just won't be tested for them.
> 
>     * I am not sure why third type of adjustment in `MiniSnippetsSessionJump` is needed. Was there a particular reason for it to be added?

Thank you very much, it worked with `config.expand.insert` instead of `MiniSnippetsSessionStart`.

>     * I am not sure why third type of adjustment in `MiniSnippetsSessionJump` is needed. Was there a particular reason for it to be added?

I was trying to prevent text from being overwritten or reset when using `jumpPre` or jump with `noinsert`, but you already gave me a solution, and the `config.expand.insert` with `noselect` worked. That was it.

I noticed that when I change the choices, there are three options like "one", "two", "three". Sometimes, when I press `<C-n>`, it selects "three" twice, and in the next one, it behaves as if there were four options. This is probably caused by `noinsert` or `noselect`, but still, thank you very much, it helped me.

**Comment by echasnovski** [comment]
> I copied `H.get_default_context` and modified the return value to be only `lang`.

Please, don't do that. There is a [documented way](https://github.com/echasnovski/mini.nvim/blob/546b9c8bd8ebba594bd18811e54bd3e6f9a662ec/doc/mini-snippets.txt#L890) of getting default context.

> Is there a reason you are using the `buf_id/lang` combination?

Yes, because 'vim.b.minisnippets_config` can contain buffer-local snippets. And that *should* be respected.

But that is not the biggest issue here as users can use their own context values for a more granular evaluation of loaders. So this type of caching can be actively bad.

**Comment by krovuxdev** [comment]
@echasnovski, I have a question: is there a way to do something like `GIT:USERNAME` to get the git username? 

But neither of the two options works, only `GIT_USERNAME` works.

```lua
lookup = {
            GIT_USERNAME = vim.fn.system("git config user.name"), -- works
            -- ["GIT:USERNAME"] = vim.fn.system("git config user.name"), doesn't work
            GIT = {
                USERNAME = vim.fn.system("git config user.name"),
            },
        },
```
global.json
```json
         "name": {
        "prefix": "git_username",
        "body": [
            "user: ${GIT:USERNAME}"
        ],
        "description": " git username"
    }
   ```
(I like the structure where `GIT = {USERNAME, EMAIL, ETC}`. It’s very convenient)

**Comment by echasnovski** [comment]
> @echasnovski, I have a question: is there a way to do something like `GIT:USERNAME` to get the git username?

No, it is not possible. Variable names must match the `[_a-zA-Z] [_a-zA-Z0-9]*` regular expression (i.e. start with Latin letter or underscore and contain only Latien letters, digits, or underscore), which is described in [specification grammar](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#snippet_syntax).

Nested lookup is also not allowed, as [documented](https://github.com/echasnovski/mini.nvim/blob/546b9c8bd8ebba594bd18811e54bd3e6f9a662ec/doc/mini-snippets.txt#L1182).

**Comment by echasnovski** [comment]
> Seems I've found a bug. Consider a following snippet for Python:
> 
> ```lua
> return {
>   ["'try/except' clause"] = {
>     prefix = "te",
>     body = {
>       "try:",
>       "\t${0:$TM_SELECTED_TEXT}",
>       "except Exception:",
>       "\tpass",
>     },
>   },
> }
> ```
> ...
> The problem is that in case of multiline `$TM_SELECTED_TEXT` only first line is correctly indented so I have to reindent.

Variables now should preserve relative indent on `main` branch. Thankfully, the actual implementation was relatively straightforward thanks to one small decision in `parse()` design which I contemplated for several days (to preserve all data about nodes during normalization and just add `text` field or collapse them into text nodes with fewer nodes in total).

The `$TM_SELECTED_TEXT` now also works better with linewise selections (no extra new line at the end; this seems to be more useful in practice).

I postponed the same change for linked tabstops (like `$1\n\t$1` and `$1\n${2:\t$1}`), as it is not clear if it is a good idea and implementation also might be involved.

**Comment by echasnovski** [comment]
I've pushed an updated for linked tabstops to preserve relative indent. It is easier to demonstrate what it means:

https://github.com/user-attachments/assets/92c3fc5a-a91d-4be0-99ea-7a7fd45befae

I have a bad feeling that this might not account for some weird edge cases or that it is not wanted as a feature. But it seems natural to handle tabstop text in the same way as variable's text (which is indeed useful).

Thanks again, @9seconds, for the general idea of "preserve relative indent".

**Comment by AlexKurisu** [comment]
Hello, and thank you for another amazing plugin! I found the following compatibility issue between `mini.snippets` and https://github.com/rafamadriz/friendly-snippets: snippet `#guard` for C language gets incorrectly expanded to
```c
#ifndef INCLUDE/home/xshell/Stuff/Sources/dev/precis-c/srcdatadata.h_
#define INCLUDE/home/xshell/Stuff/Sources/dev/precis-c/srcdatadata.h_

#endif  // INCLUDE/home/xshell/Stuff/Sources/dev/precis-c/srcdatadata.h_
```
LuaSnip expands the same snippet (correctly) to
```c
#ifndef INCLUDE_SRC_DATA_H_
#define INCLUDE_SRC_DATA_H_

#endif  // INCLUDE_SRC_DATA_H_

```

**Comment by echasnovski** [comment]
> snippet `#guard` for C language gets incorrectly expanded to

That snippet [involves variable transform](https://github.com/rafamadriz/friendly-snippets/blob/efff286dd74c22f731cdec26a70b46e5b203c619/snippets/cpp/cpp.json#L272-L277) which ['mini.snippets' doesn't support](https://github.com/echasnovski/mini.nvim/blob/bac6c8bb77fe0a872719ea43c39e35c7c695f05e/doc/mini-snippets.txt#L51-L53). LuaSnip has this functionality [through third party dependencies](https://github.com/L3MON4D3/LuaSnip/tree/master/deps) which I don't want to bring to 'mini.snippets'. Realistically, the only way 'mini.snippets' would support variable and tabstop transformations is if Neovim itself exports this kind of functionality (maybe through `vim.snippet`).

**Comment by mehalter** [comment]
I think I have encountered a bug when using `mini.snippets` while having `showmode` turned off.

Steps to reproduce:

1. Save the configuration provided below as `repro.lua` into a dedicated folder (the sandbox environment will be contained within there)
2. run `nvim -u repro.lua test.lua` from within the folder to edit some new lua file
4. Go into insert mode, type `for` and press `<C-j>`
5. Get an error, "Pattern not found"

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system "curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua")()

-- set `showmode` to false
vim.opt.showmode = false

require("lazy.minit").repro {
  spec = {
    -- basic quick start for `mini.snippets`
    {
      "echasnovski/mini.snippets",
      dependencies = "rafamadriz/friendly-snippets",
      opts = function(_, opts)
        local gen_loader = require("mini.snippets").gen_loader
        opts.snippets = {
          gen_loader.from_lang(),
        }
      end,
    },
  },
}
```

**Comment by echasnovski** [comment]
> I think I have encountered a bug when using `mini.snippets` while having `showmode` turned off.
> 
> Steps to reproduce:

Thanks! Yeah, I can reproduce. The reason is not only a 'noshowmode' (as I daily drive it) but also an absence of 'c' flag in 'shortmess' (which I have present in the option). Luckily the fix seems to be relatively straightforward and should be already be present on latest `main`.

The small side effect is that it temporarily hides '--INSERT--' on start and after jumps in case 'showmode' *is* enabled, but that seems like a reasonable compromise. Besides, 'mini.nvim' "promotes" 'noshowmode' as there is mode information in 'mini.statusline'.

**Comment by mehalter** [comment]
Thanks so much for the fix! I also just encountered a new one in terms of `mini.snippets` being unable to detect that the snippet has been deleted and therefore stopping the snippet session correctly. (this can use the same minimal reproducing config I shared before)

If you have the file `test.lua`:

```lua
local x = "top"
local y = "bottom"
```

Steps to Reproduce:

1. `nvim -u repro.lua test.lua`, open the test file with a minimal `mini.snippets` configuration
2. open a new line between the variables, if your cursor is on the first line, press `o`
3. type `for` and press `<C-j>` to complete
4. Press `<Esc>` to leave insert mode and delete the new lines, `3dd`
5. See that the session hasn't closed out and the virtual text for the body of the `for` loop is incorrectly left
![2025-01-09_10:12:35_screenshot](https://github.com/user-attachments/assets/6cf33cc6-3c3f-43a5-a77f-a44bc9e6c32b)

**Comment by echasnovski** [comment]
> I also just encountered a new one in terms of `mini.snippets` being unable to detect that the snippet has been deleted and therefore not stopping the snippet session correctly. (this can use the same minimal reproducing config I shared before)

This is expected and [is suggested to not be done](https://github.com/echasnovski/mini.nvim/blob/890db26e37bb28d8a95f05570dd0d9fa757b698b/doc/mini-snippets.txt#L1067-L1070). Snippet session is stopped automatically only [under these conditions](https://github.com/echasnovski/mini.nvim/blob/890db26e37bb28d8a95f05570dd0d9fa757b698b/doc/mini-snippets.txt#L1051-L1059). Deleting the line with tabstop's extmark is not the condition to stop session.

I tried to make it happen as it indeed seems reasonable, but there is a fundamental issue of detecting such cases here. It is reasonable to allow text editing outside of snippet session, and thus extmarks are allowed to change their position as they like. Using `invalidate = true` for extmarks (designed for this kind of scenarios) also did not seem possible because I want users to be able to remove all tabstop text and *still* preserve extmark's visualizations; `invalidate = true` does not allow that.

**Comment by mehalter** [comment]
Yeah that makes total sense! Thanks for the detailed explanation. Do you think it could make sense to add a configuration option to make the exit to normal mode more aggressive? Like have an option to just make leaving into normal mode always close the session rather than only if the final tabstop is focused?

**Comment by echasnovski** [comment]
> Do you think it could make sense to add a configuration option to make the exit to normal mode more aggressive? Like have an option to just make leaving into normal mode always close the session rather than only if the final tabstop is focused?

I don't think it is a good idea to have a dedicate option for this. Mostly because there are several ways to handle this without obvious favorite.

This can be done manually in several ways:
- Custom `ModeChanged *:n` autocommand that uses [this snippet](https://github.com/echasnovski/mini.nvim/blob/890db26e37bb28d8a95f05570dd0d9fa757b698b/doc/mini-snippets.txt#L1163-L1165) to stop all active sessions. This is probably what most users expect (so maybe deserves a special example in help, not sure).
- Other two ways described in [this comment](https://github.com/echasnovski/mini.nvim/issues/1428#issuecomment-2563668092). The first (use `<Esc>` to stop session) is the easiest and will "prevent" going into Normal mode at first press requiring second one. The second one is the most complicated and is a more performant version of "global" `ModeChanged` autocommand. If anything, *this* code should be an example, but it is too big.

**Comment by mcauley-penney** [comment]
I've checked the documentation and the source code and cannot find a way to replace the default placeholder characters, the bullet and square. Is this supported and, if not, is this a feature that will be supported in the future?

**Comment by echasnovski** [comment]
> I've checked the documentation and the source code and cannot find a way to replace the default placeholder characters, the bullet and square, with different characters. Is this supported and, if not, is this a feature that will be supported in the future?

It is done by customizing `config.expand.insert`: create custom function which calls `default_insert()` with proper options. Here is [the example](https://github.com/echasnovski/mini.nvim/blob/890db26e37bb28d8a95f05570dd0d9fa757b698b/doc/mini-snippets.txt#L683-L685) and here are [relevant options](https://github.com/echasnovski/mini.nvim/blob/890db26e37bb28d8a95f05570dd0d9fa757b698b/doc/mini-snippets.txt#L1091-L1094).

**Comment by mcauley-penney** [comment]
Thank you, I apologize for missing that

**Comment by echasnovski** [comment]
@mehalter, there is now [an example in docs](https://github.com/echasnovski/mini.nvim/blob/3a3178419ce9947f55708966dabf030eca40735a/doc/mini-snippets.txt#L451-L466) about how to set up "stop all session on Normal mode exit". Technically, it will create several 'ModeChanged' autocommands in case of nested sessions. But it still works and accounting for that will add more (unnecessary) lines to the example which I try to avoid.

**Comment by mehalter** [comment]
That's great! The example looks great! I agree with keeping the example simple and people who want to build on it have a good base to do so.

**Comment by echasnovski** [comment]
With the 0.15.0 release, 'mini.snippets' is now out of beta-testing. Thank you all for the feedback!

**Comment by aweis89** [comment]
I noticed that kubernetes snippets don't load automatically when I open a yaml file, how would I go about configuring that behavior?
